tests/python_package_test/test_engine.py

```diff
@@ -1417,42 +1416,150 @@ def test_get_split_value_histogram(self):
+                'verbose': -1,
+                'seed': 123
+            }
+            gbm = lgb.train(dict(params, first_metric_only=first_metric_only), lgb_train,
+                            num_boost_round=25, valid_sets=[lgb_eval],
+                            early_stopping_rounds=5, verbose_eval=False)
+            self.assertEqual(gbm.best_iteration, assumed_iteration)
+
+        def metrics_combination_cv_regression(metric_list, assumed_iteration,
+                                              first_metric_only, eval_train_metric):
+            params = {
+                'objective': 'regression',
+                'learning_rate': 0.9,
+                'num_leaves': 10,
+                'metric': metric_list,
+                'verbose': -1,
+                'seed': 123,
+                'gpu_use_dp': True
+            }
+            ret = lgb.cv(dict(params, first_metric_only=first_metric_only),
+                         stratified=False,
+                         train_set=lgb_train,
+                         num_boost_round=25,
+                         early_stopping_rounds=5, verbose_eval=False,
+                         eval_train_metric=eval_train_metric)
+            self.assertEqual(len(ret[list(ret.keys())[0]]), assumed_iteration)
+
+        best_iter_l1 = 16
+        best_iter_l2 = 19
+        best_iter_min = min([best_iter_l1, best_iter_l2])
+        metrics_combination_train_regression('l2', best_iter_l2, True)
+        metrics_combination_train_regression('l1', best_iter_l1, True)
+        metrics_combination_train_regression(['l2', 'l1'], best_iter_l2, True)
+        metrics_combination_train_regression(['l1', 'l2'], best_iter_l1, True)
+        metrics_combination_train_regression(['l2', 'l1'], best_iter_min, False)
+        metrics_combination_train_regression(['l1', 'l2'], best_iter_min, False)
+
+        best_iter_l1 = 6
+        best_iter_l2 = 11
+        best_iter_min = min([best_iter_l1, best_iter_l2])
+        metrics_combination_cv_regression('l2', best_iter_l2, True, False)
+        metrics_combination_cv_regression('l1', best_iter_l1, True, False)
+        metrics_combination_cv_regression(['l2', 'l1'], best_iter_l2, True, False)
+        metrics_combination_cv_regression(['l1', 'l2'], best_iter_l1, True, False)
+        metrics_combination_cv_regression(['l2', 'l1'], best_iter_min, False, False)
+        metrics_combination_cv_regression(['l1', 'l2'], best_iter_min, False, False)
+        metrics_combination_cv_regression('l2', best_iter_l2, True, True)
+        metrics_combination_cv_regression('l1', best_iter_l1, True, True)
+        metrics_combination_cv_regression(['l2', 'l1'], best_iter_l2, True, True)
+        metrics_combination_cv_regression(['l1', 'l2'], best_iter_l1, True, True)
+        metrics_combination_cv_regression(['l2', 'l1'], best_iter_min, False, True)
+        metrics_combination_cv_regression(['l1', 'l2'], best_iter_min, False, True)
+
+        # classification test
```

Comment:
Maybe it's enough to test only regression? It seems that there is no differentiation for regression and classification in `first_metric_only` mechanism implementation. I mean, **regression test** and **classification test** check the same things, just for different metrics...

Reply:
If you'd like to remove classification test for simplification, I agree with you since there may be no difference in the light of early stopping specification. So I committed to delete the test.

Reply:
> Also, it seems to me that something is wrong with sklearn wrapper (first_metric_only=True disables early stopping completely (?)). Can you please check it?

I haven't realized the behavior, so I'm going to investigate it. Please wait for a while.